src/coreclr/src/vm/eventpipesession.cpp

```diff
@@ -482,12 +537,32 @@ void EventPipeSession::SuspendWriteEvent()
         THROWS;
         GC_TRIGGERS;
         MODE_PREEMPTIVE;
+        // Need to disable the session before calling this method
+        PRECONDITION(!EventPipe::IsSessionEnabled(reinterpret_cast<EventPipeSessionID>(this)));
     }
     CONTRACTL_END;
 
-    // Force all in-progress writes to either finish or cancel
-    // This is required to ensure we can safely flush and delete the buffers
-    m_pBufferManager->SuspendWriteEvent(GetIndex());
+    {
+        SpinLockHolder holder(EventPipeThread::GetGlobalThreadLock());
+
+        EventPipeThreadIterator eventPipeThreads = EventPipeThread::GetThreads();
+        EventPipeThread *pThread = nullptr;
+        while (eventPipeThreads.Next(&pThread))
+        {
+            // Wait for the thread to finish any writes to this session
+            YIELD_WHILE(pThread->GetSessionWriteInProgress() == GetIndex());
+
+            // Since we've already disabled the session, the thread won't call back in to this
+            // session once its done with the current write
+        }
+    }
+
+    if (m_pBufferManager != nullptr)
+    {
+        // Force all in-progress writes to either finish or cancel
```

Comment:
This comment is no longer accurate, in progress writes finished above. All this does now is convert the thread write buffers into the read-only state ensuring they will be flushed in WriteAllBuffersToFile() later.